book.md

```diff
@@ -495,17 +492,160 @@ probably include, or be followed by, a section on the various sanitizers
+
+In the early days of memory vulnerability exploitation, attackers could simply
+place shellcode\index{shellcode} of their choice in executable memory and jump
+to it. As executable-space protection became mainstream, attackers started to
+reuse code already present in an application's binary and linked libraries
+instead. A variety of different techniques to this effect came to light.
+
+The simplest of these techniques is return-to-libc [@Solar1997]. Instead of
+returning to shellcode that the attacker has injected, the return address is
+modified to return into a library function, such as `system` or `exec`. This
+technique is simpler to use when arguments are also passed on the stack and
+can therefore be controlled with the same stack buffer overflow that is used
+to modify the address.
+
+Return-to-libc attacks restrict an attacker to whole library functions. While
+this can lead to powerful attacks, it has also been demonstrated that it is
+possible to achieve arbitrary computation by combining a number of short
+instruction sequences known as **gadgets**\index{gadget}. In return-oriented
+programming (ROP)\index{return-oriented programming (ROP)} [@Shacham2007], In
+ROP, each gadget performs a simple operation, for example setting a register,
+then pops a return address from the stack and returns to it. The attacker
+constructs a fake call stack (often called a ROP chain) which ensures a number
+of gadgets are executed one after another, in order to perform a more complex
+operation.
+
+This will hopefully become more clear with an example: a ROP chain for
+AArch64 that starts a shell, by calling `execve` with `"/bin/sh"` as an
+argument. The prototype of the `execve` library function, which wraps the
+exec system call , is:
+
+```
+  int execve(const char *pathname, char *const argv[],
+             char *const envp[]);
+```
+For AArch64, `pathname` will be passed in the `x0` register, `argv` will be
+passed in `x1`, and `envp` in `x2`. For starting a shell, it is sufficient
+to:
+
+ * Make `x0` contain a pointer to `"/bin/sh"`.
+ * Make `x1` contain a pointer to an array of pointers with two elements:
+   * The first element is a pointer to `"/bin/sh"`.
+   * The second element is zero (`NULL`).
+ * Make `x2` contain zero (`NULL`).
+
+This can be achieved by chaining gadgets to set the registers `x0`, `x1`,
+`x2`, and then returning to `execve` in the C library.
+
+Let's assume we have the following gadgets:
+
+  1. A gadget that loads `x0` and `x1` from the stack:
+```
+  gadget_x0_x1:
+    ldp x0, x1, [sp]
+    ldp x20, x19, [sp, #64]
+    ldp x29, x30, [sp, #32]
+    ldr x21, [sp, #48]
+    add sp, sp, #0x50
+    ret
+```
+
+ 2. A gadget that sets `x2` to zero, but also clears `x0` as a side-effect:
+```
+  gadget_x2:
+    mov x2, xzr
+    mov x0, x2
+    ldp x20, x19, [sp, #32]
+    ldp x29, x30, [sp]
+    ldr x21, [sp, #16]
+    add sp, sp, #0x30
+    ret
+```
```

Comment:
When reading the gadget assembly, I was wondering if these gadgets could easily emerge from normal source code. If it's not explained later, and you'd happen to know that these gadgets can be generated easily from e.g. C code, maybe it'd be worthwhile to also show in the book which C code produced these gadgets?

Reply:
Adding a TODO for this.